pkg/util/pods.go

```diff
@@ -134,79 +138,68 @@ func WaitForPodCompletion(ctx context.Context, podClient kubernetes.PodClient, n
 }
 
 func waitForPodCompletionOrTimeout(ctx context.Context, podClient kubernetes.PodClient, namespace, name string, completed map[string]time.Time, notifier ContainerNotifier, skipLogs bool) (*corev1.Pod, error) {
-	// Warning: this is extremely fragile, inherited legacy code.  Please be
-	// careful and test thoroughly when making changes, as they very frequently
-	// lead to systemic production failures.  Some guidance:
-	// - There is a complex interaction between this code and the container
-	//   notifier.  Updates to the state of the pod are received via the watch
-	//   and communicated to the notifier.  Even in case of interruption (i.e.
-	//   cancellation of `ctx`) and/or failure, events should continue to be
-	//   processed until the notifier signals that it is done.  This ensures
-	//   the state of the pod is correctly reported, artifacts are gathered,
-	//   and termination happens deterministically for both success and failure
-	//   scenarios.
-	// - Since ea8f62fcf, most of the above only applies to template tests.
-	//   Container and multi-stage tests now solely rely on `test-infra`'s
-	//   `pod-utils` for artifact gathering and so use a notifier which
-	//   instantly reports itself as done when the watched containers finish.
-	pod := &corev1.Pod{}
-	if err := podClient.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: namespace, Name: name}, pod); err != nil {
-		if kerrors.IsNotFound(err) {
-			notifier.Complete(name)
-			logrus.Infof("error: could not wait for pod '%s': it is no longer present on the cluster"+
-				" (usually a result of a race or resource pressure. re-running the job should help)", name)
-			return nil, fmt.Errorf("pod was deleted while ci-operator step was waiting for it")
+	var ret atomic.Pointer[corev1.Pod]
```

Comment:
Nice that golang has it now :+1:

Reply:
Finally!  Even though [getting there](https://github.com/openshift/ci-tools/pull/3282) wasn't easy =)